File: internal/redis/token_bucket.lua
-- KEYS[1] = rate limit key
-- ARGV[1] = now (unix nanos)
-- ARGV[2] = capacity
-- ARGV[3] = refill_rate (tokens per second)

local key = KEYS[1]
local now = tonumber(ARGV[1])
local capacity = tonumber(ARGV[2])
local refill_rate = tonumber(ARGV[3])

local data = redis.call("HMGET", key, "tokens", "last_refill")

local tokens = tonumber(data[1])
local last_refill = tonumber(data[2])

if tokens == nil then
    tokens = capacity
    last_refill = now
end

local elapsed = now - last_refill
if elapsed > 0 then
    local refill = math.floor(elapsed * refill_rate / 1000000000)
    if refill > 0 then
        tokens = math.min(capacity, tokens + refill)
        last_refill = now
    end
end

if tokens <= 0 then
    redis.call("HMSET", key,
        "tokens", tokens,
        "last_refill", last_refill
    )
    return {0, tokens}
end

tokens = tokens - 1

redis.call("HMSET", key,
    "tokens", tokens,
    "last_refill", last_refill
)

local ttl_seconds = 60

if tokens <= 0 then
    redis.call("HMSET", key,
        "tokens", tokens,
        "last_refill", last_refill
    )
    redis.call("EXPIRE", key, ttl_seconds)
    return {0, tokens}
end

tokens = tokens - 1

redis.call("HMSET", key,
    "tokens", tokens,
    "last_refill", last_refill
)
redis.call("EXPIRE", key, ttl_seconds)

return {1, tokens}

File: internal/fastpath/tokenBucket.go
package fastpath

import "time"

type TokenBucket struct {
	capacity   int64
	tokens     int64
	refillRate int64 // tokens per second
	lastRefill int64 // unix nanos
}

func NewTokenBucket(capacity, refillRate int64, now int64) *TokenBucket {
	return &TokenBucket{
		capacity:   capacity,
		tokens:     capacity,
		refillRate: refillRate,
		lastRefill: now,
	}
}

func (b *TokenBucket) Allow(now int64) bool {
	elapsed := now - b.lastRefill
	if elapsed > 0 {
		refill := (elapsed * b.refillRate) / int64(time.Second)
		if refill > 0 {
			b.tokens += refill
			if b.tokens > b.capacity {
				b.tokens = b.capacity
			}
			b.lastRefill = now
		}
	}

	if b.tokens <= 0 {
		return false
	}

	b.tokens--
	return true
}

func (b *TokenBucket) RefillRate() int64 {
	return b.refillRate
}

func (b *TokenBucket) Tokens() int64 {
	return b.tokens
}
